### PR TITLE
Gutenberg: Add the Calypso Media Modal

### LINKED ISSUE #26904: Gutenberg: Media Modal
When manipulating any media let's make sure the Calypso Media Modal can be used. Let's write up additional tasks if we find any blockers or specific issues to work through.

See also https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/components/media-upload/index.js

Let's spike on this for 2 days and report back on how much work we think this is. (Note that we're say basic functionality, no image editor, no customizations like other image search).

### LINKED ISSUE #27641: Gutenberg: Inline Image
Inline image blocks do not appear to work

### PR DESCRIPTION
**Pull request by Copons**
#### Changes proposed in this Pull Request

* Add the Calypso Media Modal in Gutenberg.
* Wire it up with the `Image`, `Cover Image`, `Gallery`, `File`, and `Inline Image` blocks.

#### Notes

- Gutenberg's `Gallery` block doesn't seem to support the same options as Calypso gallery shortcode (e.g. the layout), but it simply has a slider for the number of columns.
This is why I've decided, for now, to disable the edit gallery view in the media modal.

- For some reasons, when opening the modal through the `Inline Image` block, the `onOpen` function is called _a lot_ of times. The `setState` contained inside is called as many times, to the point that it crashes the modal.
I couldn't figure out why (in Core it is called only once, but then again, the logic is very different), but I've worked around this issue by [only calling `setState` if needed](https://github.com/Automattic/wp-calypso/pull/27687/files#diff-da6effaf543f6ec6542895c04d32ce4cR42).

- Modifying media in the modal automatically updates the blocks using those media by listening to `MediaStore.on( 'change' )`.
Currently though, `Inline Image` blocks are not updated accordingly and, apparently, if their image has been changed, they will be removed on reload.

- `Audio` and `Video` blocks will be handled in one or more follow up PRs.

#### Testing instructions

* Open Gutenberg in Calypso dev: http://calypso.localhost:3000/gutenberg/post/
* Add an `Image` block.
* Click on the Media Library button.
* Make sure the Media Modal shows up, and the images are loaded as expected.
* Try inserting one image and make sure it's indeed added to the block.
* Try reloading the editor and check if the `Image` block shows the added image.
* Try editing the block and check that the image is selected in the Media Library.
* Do the same tests for `Cover Image`, `Gallery`, and `File`.
* Move the cursor inside a Paragraph block and try inserting an `Inline Image`. Make sure the media modal shows up, it's possible to insert an inline image in the paragraph, and that the block still works after a save and reload.

Fixes #26904
Fixes #27641

### CODE REVIEW

File: client/gutenberg/editor/edit-post/hooks/components/media-upload/index.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
+/** @format */
```
Comment (vindl) [inline client/gutenberg/editor/edit-post/hooks/components/media-upload/index.js]: Nitpick: same thing I pointed out in some of the previous PRs - if possible I'd like to keep all of our custom additions outside of `edit-post` folder for now.

Ideally, at one point I'd like to replace all of it by `edit-post` package, and deleting the whole folder would be much trickier if he have Calypso's custom additions all over it. Let me know if that makes sense?

**Comment by Copons** [comment]
> Just a note that in this case the insert button is not disabled while uploading (based on spinner at least) and I was able to insert the image before the upload was completed. However, everything worked out as expected and it's probably a better/faster flow than blocking the insert for upload.

Ah good catch, that line was a leftover from a previous version.
Since then, I've replaced the blocker with the MediaStore listener that refreshes the blocks upon media change.

File: client/gutenberg/editor/hooks/components/media-upload/utils.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		sizes: {
+			full: {
+				url: get( media, 'URL' ),
+			},
+			...reduce(
+				media.thumbnails,
+				( thumbnails, url, size ) => {
+					thumbnails[ size ] = { url };
+					return thumbnails;
+				},
```
Comment (dmsnell) [inline client/gutenberg/editor/hooks/components/media-upload/utils.js]: also known as…

```js
( thumbnails, url, size ) => ( { ...thumbnails, [ size ]: { url } } )
```

just a fun fact!

Comment (Copons) [reply client/gutenberg/editor/hooks/components/media-upload/utils.js]: It's not _that_ fun! 😤

**Pull request merged**